src/bert_layers/attention.py

```diff
@@ -857,7 +966,15 @@ def forward(
             qkv = self.rotary_emb(qkv, seqlen_offset=seqlen_offset, max_seqlen=None)
 
             q, k, v = qkv.transpose(3, 1).unbind(dim=2)  # b h s d
-            attn = F.scaled_dot_product_attention(q, k, v, dropout_p=self.p_dropout)
+            attn = F.scaled_dot_product_attention(
+                q,
+                k,
+                v,
+                dropout_p=self.p_dropout,
+                attn_mask=attn_mask[:, None, None, :seqlen].to(torch.bool).expand(bs, 1, seqlen, seqlen)
```

Comment:
`bs` should be `unpad_bs` (took some time to understand, BS is really misleading for unpadded inputs)